### DISCUSSION TITLE
Milvus querynodes out of memory issue

### DISCUSSION THREAD

**Discussion by smstoyanov** [milvus-io/milvus#35713] [Q&A and General discussion] [answered]
Hello, I have deployed milvus on kubernetes using helm chart version: 4.1.28 .  The storage backend is s3 and  our query nodes are crashlooping with OOM after long run test with about 1000 concurrent connections. The resources we request for the query nodes are 16G and 2CPUs. 
The values config of the installation is:
`milvus:
  nameOverride: "milvus"
  fullnameOverride: "milvus"
  serviceAccount:
    create: true
    name: "milvus"
    annotations:
      eks.amazonaws.com/role-arn: "${milvus-iam-role-arn}"

  cluster:
    enabled: true

  service:
    type: ClusterIP
    port: 19530

  minio:
    enabled: false

  externalS3:
    enabled: true
    host: "s3.${aws-region}.amazonaws.com"
    port: "443"
    useSSL: true
    bucketName: "${my_aws_bucket}"
    useIAM: true
    cloudProvider: "aws"
    iamEndpoint: "iam.amazonaws.com"
    region: "${aws-region}"

  rootCoordinator:
    replicas: 2
    activeStandby:
      enabled: true
    resources:
      limits:
        cpu: 1
        memory: 2Gi

  indexCoordinator:
    replicas: 2
    activeStandby:
      enabled: true
    resources:
      limits:
        cpu: "0.5"
        memory: 0.5Gi

  queryCoordinator:
    replicas: 2
    activeStandby:
      enabled: true
    resources:
      limits:
        cpu: "0.5"
        memory: 0.5Gi

  dataCoordinator:
    replicas: 2
    activeStandby:
      enabled: true
    resources:
      limits:
        cpu: "0.5"
        memory: 0.5Gi

  queryNode:
    replicas: 2
    resources:
      limits:
        memory: 16Gi
        cpu: 2
    nodeSelector:
      milvus: "true"

  proxy:
    replicas: 2
    resources:
      limits:
        cpu: 1
        memory: 2Gi
  pulsar:
    enabled: false
  kafka:
    enabled: true`

The tests are creating a collection with a vector field of size 1024. The index used is FLAT for 100% recall. The load test is  with 1000-1500 concurrent users and we get around 10-15 rps on average for around 1 hour. During this time run around 25K inserts and 20K queries and the resulting size of the collection is around 2.5 mil. records in the end. At some point if we run the test again we saw  that  the used memory on each query-node continue to increase  and the query-nodes Crashlooping with OOM . What we can do  to optimize our setup for our expected data target which is about 50mil records? How much resources querynodes will need to keep working smoothly even  with a little bit lower RPS  - not lower than 7.

**Comment by xiaofan-luan** [comment]
1. You need arount 500GB to put all the 50million vector in memory. Using Mmap can potentially save you some cost.
2.  You need an ann index to increase throughput. If use purely brute force search you may need huge amount of resource to run 10QPS on. 50 million records.
3. For 2.5 million data, 2* 16GB should be good enough, I don't foresee any reason why this won't work.

**Comment by yhmo** [comment, marked as answer]
Use this tool to estimate the required resources according to your dataset: https://milvus.io/tools/sizing

For 50M 1024-dim FLAT index:
![Screenshot from 2024-08-27 11-19-35](https://github.com/user-attachments/assets/0c399648-14ff-47b9-9408-4704430c90d0)

For 2.5M 1024-dim FLAT index:
![Screenshot from 2024-08-27 11-20-11](https://github.com/user-attachments/assets/d05a1852-86b5-474f-a1d1-2b446b88ad6c)

FLAT is a brute-force search, each vector is computed. It is hard to achieve 7QPS with 50M 1024-dim vectors, perhaps you need 100+ query nodes and 10X memory resources to accomplish the goal.

**Comment by smstoyanov** [comment]
thank you very much for the detailed guide.
